### PR TITLE
feat: support chaitin-waf plugin

### PR DESCRIPTION
**Pull request by lingsamuel**
### Description

Fixes # (issue)

### Checklist

- [ ] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: docs/zh/latest/plugins/chaitin-waf.md [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+X-APISIX-CHAITIN-WAF: no
+Date: Wed, 19 Jul 2023 09:30:28 GMT
+X-Content-Type-Options: nosniff
+X-Powered-By: httpbun/3c0dc05883dd9212ac38b04705037d50b02f2596
+Server: APISIX/3.3.0
+
+404 page not found
+```
+
+当满足匹配条件时，正常请求依然可以触达：
```
Comment (soulbird) [inline docs/zh/latest/plugins/chaitin-waf.md:183]: 触达上游？

File: apisix/plugins/chaitin-waf.lua [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+            end
+        end
+    end
+
+    return match_passed, nil
+end
+
+local function get_conf(conf, metadata)
+    local t = {
+        mode = "block", -- block or monitor or off, default off
```
Comment (soulbird) [inline apisix/plugins/chaitin-waf.lua:244]: default value is ’off‘ ?

File: apisix/plugins/chaitin-waf.lua [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+    end
+    return host, port, nil
+end
+
+local function check_match(conf, ctx)
+    local match_passed = true
+
+    if conf.match then
+        for _, match in ipairs(conf.match) do
+            local exp, err = expr.new(match.vars)
```
Comment (soulbird) [inline apisix/plugins/chaitin-waf.lua:224]: Can we cache the result of expression compilation？

File: apisix/plugins/chaitin-waf.lua [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+                    },
+                },
+                required = { "host" }
+            },
+            minItems = 1,
+        },
+        config = {
+            type = "object",
+            properties = {
+                -- connect timeout, in milliseconds, integer, default 1s (1000ms)
```
Comment (soulbird) [inline apisix/plugins/chaitin-waf.lua:113]: Annotation looks unnecessary

File: apisix/plugins/chaitin-waf.lua [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+local metadata_schema = {
+    type = "object",
+    properties = {
+        nodes = {
+            type = "array",
+            items = {
+                type = "object",
+                properties = {
+                    host = {
+                        type = "string",
```
Comment (soulbird) [inline apisix/plugins/chaitin-waf.lua:98]: It is recommended to use a stricter schema definition. https://github.com/apache/apisix/blob/master/apisix/schema_def.lua#L40

File: t/plugin/chaitin-waf-reject.t [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+   only: true                  # Enable L4 proxy only without L7 proxy.
+   tcp:
+     - addr: 9100              # Set the TCP proxy listening ports.
+       tls: true
+     - addr: "127.0.0.1:9101"
+   udp:                        # Set the UDP proxy listening ports.
+     - 9200
+     - "127.0.0.1:9201"
+plugins:
+    - chaitin-waf
```
Comment (soulbird) [inline t/plugin/chaitin-waf-reject.t:52]: The plugin has been enabled by default, there is no need to add it separately

File: docs/en/latest/plugins/chaitin-waf.md [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+X-APISIX-CHAITIN-WAF-TIME: 2
+X-APISIX-CHAITIN-WAF-ACTION: reject
+X-APISIX-CHAITIN-WAF-STATUS: 403
+Server: APISIX/3.3.0
+Set-Cookie: sl-session=UdywdGL+uGS7q8xMfnJlbQ==; Domain=; Path=/; Max-Age=86400
+
+{"code": 403, "success":false, "message": "blocked by Chaitin SafeLine Web Application Firewall", "event_id": "51a268653f2c4189bfa3ec66afbcb26d"}
+```
+
+## Disable Plugin
```
Comment (kayx23) [inline docs/en/latest/plugins/chaitin-waf.md:241]:
```suggestion
## Delete Plugin
```

File: docs/en/latest/plugins/chaitin-waf.md [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+
+To disable the `chaitin-waf` plugin, remove the plugin configuration in the plugin configuration:
```
Comment (kayx23) [inline docs/en/latest/plugins/chaitin-waf.md:243]:
```suggestion
To remove the `chaitin-waf` plugin, delete the  the corresponding JSON configuration from the Plugin configuration:
```

File: docs/zh/latest/plugins/chaitin-waf.md (hunk 1 above)
Comment (monkeyDluffy6017) [reply docs/zh/latest/plugins/chaitin-waf.md]: Done

File: apisix/plugins/chaitin-waf.lua (hunk 2 above)
Comment (monkeyDluffy6017) [reply apisix/plugins/chaitin-waf.lua]: Done

File: apisix/plugins/chaitin-waf.lua (hunk 3 above)
Comment (monkeyDluffy6017) [reply apisix/plugins/chaitin-waf.lua:231]: We will do it after this pr

File: apisix/plugins/chaitin-waf.lua (hunk 4 above)
Comment (monkeyDluffy6017) [reply apisix/plugins/chaitin-waf.lua]: Done

File: apisix/plugins/chaitin-waf.lua (hunk 5 above)
Comment (monkeyDluffy6017) [reply apisix/plugins/chaitin-waf.lua:97]: Done

File: t/plugin/chaitin-waf-reject.t (hunk 6 above)
Comment (monkeyDluffy6017) [reply t/plugin/chaitin-waf-reject.t]: Done

File: apisix/plugins/chaitin-waf.lua [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+                },
+                -- idle connection timeout, in milliseconds, integer, default 60s (60000ms)
+                keepalive_timeout = {
+                    type = "integer",
+                    default = 60000
+                },
+                -- remote address from ngx.var.VARIABLE, string
+                remote_addr = {
+                    type = "string",
+                    default = "http_x_forwarded_for: 1",
```
[the remark this answers is not in the document]
Comment (monkeyDluffy6017) [reply apisix/plugins/chaitin-waf.lua]: Done

File: apisix/plugins/chaitin-waf.lua [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+                    result.event_id) .. "\n", extra_headers
+        end
+    end
+    if not ok then
+        extra_headers[HEADER_CHAITIN_WAF_STATUS] = nil
+    end
+
+    return nil, nil, extra_headers
+end
+
```
[the remark this answers is not in the document]
Comment (monkeyDluffy6017) [reply apisix/plugins/chaitin-waf.lua:349]: Done

File: docs/en/latest/plugins/chaitin-waf.md (hunk 7 above)
Comment (monkeyDluffy6017) [reply docs/en/latest/plugins/chaitin-waf.md]: Done

File: docs/en/latest/plugins/chaitin-waf.md (hunk 8 above)
Comment (monkeyDluffy6017) [reply docs/en/latest/plugins/chaitin-waf.md]: Done

File: apisix/plugins/chaitin-waf.lua (hunk 3 above)
Comment (soulbird) [reply apisix/plugins/chaitin-waf.lua:231]: Add TODO pls

Comment (monkeyDluffy6017) [reply apisix/plugins/chaitin-waf.lua:232]: Done

File: docs/zh/latest/plugins/chaitin-waf.md [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+X-APISIX-CHAITIN-WAF-TIME: 2
+X-APISIX-CHAITIN-WAF-ACTION: reject
+X-APISIX-CHAITIN-WAF-STATUS: 403
+Server: APISIX/3.3.0
+Set-Cookie: sl-session=UdywdGL+uGS7q8xMfnJlbQ==; Domain=; Path=/; Max-Age=86400
+
+{"code": 403, "success":false, "message": "blocked by Chaitin SafeLine Web Application Firewall", "event_id": "51a268653f2c4189bfa3ec66afbcb26d"}
+```
+
+## 禁用插件
```
Comment (kayx23) [inline docs/zh/latest/plugins/chaitin-waf.md:238]:
```suggestion
## 删除插件
```

File: docs/en/latest/plugins/chaitin-waf.md [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+| config                   | object        | false    |               | Configuration of the Chaitin SafeLine WAF service. When the route is not configured, the parameters configured in the metadata are used.                                                                                                                                                                                                                  |
+| config.connect_timeout   | integer       | false    |               | connect timeout, in milliseconds, integer                                                                                                                                                                                                                                                                                                                 |
```
Comment (kayx23) [inline docs/en/latest/plugins/chaitin-waf.md:93]:
```suggestion
| config.connect_timeout   | integer       | false    |               | connect timeout, in milliseconds                                                                                                                                                                                                                                                                                                                 |
```

Same for below. **Remove all `integer` from the description** as the data types have already been specified in `Type`.

File: docs/en/latest/plugins/chaitin-waf.md [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+    - timeout: Timeout for request to the WAF server
+- **X-APISIX-CHAITIN-WAF-ERROR**: Debug header. WAF error message
+- **X-APISIX-CHAITIN-WAF-TIME**: The time in milliseconds that APISIX spent interacting with WAF.
+- **X-APISIX-CHAITIN-WAF-STATUS**: The status code returned to APISIX by the WAF server.
+- **X-APISIX-CHAITIN-WAF-ACTION**: Processing result returned to APISIX by the WAF server.
+    - pass: request valid and passed
+    - reject: request rejected by WAF service
+- **X-APISIX-CHAITIN-WAF-SERVER**: Debug header. Picked WAF server.
+
+## Plugin Metadata
```
Comment (kayx23) [inline docs/en/latest/plugins/chaitin-waf.md:55]:
```
## Plugin Metadata
```

I thought the other ones say "Attributes"?

**Review by leslie-tsang — APPROVED** [review]

**Review by soulbird — APPROVED** [review]

**Pull request merged**